onmt/inputters/inputter.py

```diff
@@ -270,10 +271,10 @@ def build_dataset(fields, data_type, src, src_reader,
         use_filter_pred (bool): Whether or not to apply length filtering.
     """
 
-    dataset_classes = {
-        'text': TextDataset, 'img': ImageDataset, 'audio': AudioDataset
+    sort_keys = {
+        'text': text_sort_key, 'img': img_sort_key, 'audio': audio_sort_key
     }
-    assert data_type in dataset_classes
+    assert data_type in sort_keys
```

Comment:
Is the `data_type` parameter of `build_dataset` necessary? It is used 

1. For selecting which sort key to use (`sort_key=sort_keys['data_type']`.
2. For defining the filtering logic (`use_src_len=data_type == 'text'`)

I'm not convinced making either of these things depend on the `data_type` is a good idea in the long run. There are two data types, not one, and for any given type of source or target data I might want to sort or filter by different criteria than the format of the src data.

Reply:
I guess the alternative is passing a ``sort_key`` and ``use_src_len`` (or even ``filter_pred``) argument. At that point, it probably makes sense to remove ``build_dataset`` and explicitly call the constructor everywhere. What do you think?

Reply:
That seems reasonable.

Reply:
Done.